Sources/NIO/BSDSocketAPIWindows.swift

```diff
@@ -210,15 +247,62 @@ extension NIOBSDSocket {
     }
 
     @inline(never)
-    static func recvmsg(descriptor: CInt, msgHdr: UnsafeMutablePointer<msghdr>,
-                        flags: CInt) throws -> IOResult<size_t> {
-        fatalError("recvmsg not yet implemented on Windows")
+    static func recvmsg(socket s: NIOBSDSocket.Handle,
+                        msgHdr lpMsg: UnsafeMutablePointer<msghdr>,
+                        flags: CInt)
+            throws -> IOResult<size_t> {
+        var InBuffer = WSAID_WSARECVMSG
+        var pfnWSARecvMsg: LPFN_WSARECVMSG?
+        var cbBytesReturned: DWORD = 0
+        if WinSDK.WSAIoctl(s, SIO_GET_EXTENSION_FUNCTION_POINTER,
+                           &InBuffer, DWORD(MemoryLayout.stride(ofValue: InBuffer)),
+                           &pfnWSARecvMsg,
+                           DWORD(MemoryLayout.stride(ofValue: pfnWSARecvMsg)),
+                           &cbBytesReturned, nil, nil) == SOCKET_ERROR {
+            throw IOError(winsock: WSAGetLastError(), reason: "WSAIoctl")
+        }
+
+        guard let pfnWSARecvMsg = pfnWSARecvMsg else {
+            throw IOError(windows: DWORD(ERROR_INVALID_FUNCTION),
+                          reason: "recvmsg")
+        }
```

Comment:
So let's state up-front that this code is entirely insane. Worse, it seems to be correct, and I think having insane and correct code is the worst of all worlds.

As we noted in our Slack chat, this is a performance nightmare. Ideally we'd cache this function pointer, if we can work out what its validity period is, and we'd also deduplicate the code from the sendmsg implementation. However, for now I'm ok keeping it here, noting that it limits performance, because "working but slow" is better than "not working". Anything can be fast if it doesn't work.

That said, can we have a comment at the top of this that reminds us to find a way to dedupe this and cache the function pointer?